webapp/src/cloudsync.ts

```diff
@@ -656,34 +688,44 @@ function userInitials(username: string): string {
     return ((initials.shift() || '') + (initials.pop() || '')).toUpperCase();
 }
 
-data.mountVirtualApi("sync", {
-    getSync: p => {
-        switch (data.stripProtocol(p)) {
-            case "user":
-                const user = pxt.Util.jsonTryParse(pxt.storage.getLocal("cloudUser")) as pxt.editor.UserInfo;
-                return user;
-            case "loggedin":
-                return currentProvider != null
-            case "status":
-                return status
-            case "hascloud":
-                return providers().length > 0
-            case "hassync":
-                return currentProvider && currentProvider.hasSync();
-            case "provider":
-                return currentProvider && currentProvider.name;
-            case "providericon":
-                if (currentProvider)
-                    return currentProvider.icon;
-                const prs = providers();
-                if (prs.length == 1)
-                    return prs[0].icon;
-                return "user";
-        }
-        return null
-    },
-})
+function syncApiHandler(p: string) {
```

Comment:
This seems like the main state machine in the auth flow. Is that correct? Would you mind providing some code comments here or elsewhere in the file giving a high level picture of what's happening?

Some questions:
- How does the data flow with the providers work? e.g. 1st, hit this github API, 2nd pop this dialog to get consent, etc.
- What do we store locally and why?
- When and why do we interact with the backend?
- Any gotcha's or non-standard stuff we do?